src/repl-frontend/scala/tools/nsc/interpreter/jline/Reader.scala

```diff
@@ -0,0 +1,223 @@
+/*
+ * Scala (https://www.scala-lang.org)
+ *
+ * Copyright EPFL and Lightbend, Inc.
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.tools.nsc.interpreter
+package jline
+
+import shell.{Accumulator, ShellConfig}
+
+import org.jline.reader.{
+  Candidate, Completer, CompletingParsedLine,
+  EndOfFileException, UserInterruptException, EOFError,
+  History, LineReader, ParsedLine, Parser}, Parser.ParseContext
+import org.jline.reader.impl.DefaultParser
+
+import java.util.{List => JList}
+
+/** A Reader that delegates to JLine3.
+ */
+class Reader private (config: ShellConfig, reader: LineReader, val accumulator: Accumulator, val completion: shell.Completion) extends shell.InteractiveReader {
+  val history: shell.History = new HistoryAdaptor(reader.getHistory)
+  def interactive: Boolean = true
+  protected def readOneKey(prompt: String): Int = ???
+  protected def readOneLine(prompt: String): String = {
+    try {
+      reader.readLine(prompt)
+    } catch {
+      case _: EndOfFileException | _: UserInterruptException => reader.getBuffer.delete() ; null
+    }
+  }
+  def redrawLine(): Unit = ???
+  def reset(): Unit = accumulator.reset()
+}
+
+object Reader {
+  import org.jline.reader.LineReaderBuilder
+  import org.jline.reader.impl.history.DefaultHistory
+  import org.jline.terminal.TerminalBuilder
+
+  /** Construct a Reader with various JLine3-specific set-up.
+   *  The `shell.Completion` is wrapped in the `jline.Completion` bridge to enable completion from JLine3.
+   */
+  def apply(config: ShellConfig, repl: Repl, completion: shell.Completion, accumulator: Accumulator): Reader = {
+    require(repl != null)
+    if (config.isReplDebug) initLogging()
+
+    System.setProperty(LineReader.PROP_SUPPORT_PARSEDLINE, java.lang.Boolean.TRUE.toString())
+
+    //val terminal = TerminalBuilder.builder().build()
+    val terminal  = TerminalBuilder.terminal() // defaults for now
+    val completer = new Completion(completion)
+    val parser    = new ReplParser(repl)
+    val history   = new DefaultHistory
+
+    val builder =
+      LineReaderBuilder.builder()
+      .appName("scala")
+      .completer(completer)
+      .history(history)
+      .parser(parser)
+      .terminal(terminal)
```

Comment:
In our usage of JLine2, we picked up configuration from `~/inputrc` [etc](https://github.com/jline/jline2/blob/12b98d94589e3bd6a60719c495b0ab911c571329/src/main/java/jline/console/ConsoleReader.java#L290-L306).

JLine3 does have support for parsing this file since https://github.com/jline/jline3/commit/9ccfe0b121d2ea72a8524c84a9bda5abaa68d45a, but it looks like we need to [call `InputRC.configure`](https://github.com/jline/jline3/commit/9ccfe0b121d2ea72a8524c84a9bda5abaa68d45a#diff-b26d899b1659f142e858937634634a4bR56) after determining which config file to use ourselves.

Furthermore, in the current configuration file, sections guarded with `$if JLine .... $endif` are interpreted by the Scala REPL. Under this change, config would need to be moved to `$if scala` (the `appname` used here. That's probably fine if release noted.

Here's my `.inputrc` for reference.

```
$if Bash
	"\e[A": history-search-backward
	"\e[B": history-search-forward
    Space: magic-space
$endif

#set blink-matching-paren on
"\C-xq": "\eb\"\ef\""

"\C-x)": "\ea\(\ee\)"

$if JLine
	"\C-xu": "import scala.tools.reflect.ToolBox; val m = scala.reflect.runtime.currentMirror; val u = m.universe; import u._; val tb = m.mkToolBox()"
	"\C-x)": "\C-a(\C-e)"
	"\C-x}": "\C-a{\C-e}"
	"\C-x{": "{}\C-b"
	"\C-x(": "()\C-b"

$endif
```

Reply:
spun out as https://github.com/scala/scala-dev/issues/683